src/extension_commands.ml

```diff
@@ -950,6 +950,220 @@ module Search_by_type = struct
   ;;
 end
 
+module Navigate_holes = struct
+  let extension_name = "Navigate between Typed Holes"
+
+  let ocaml_lsp_doesnt_support_typed_holes ocaml_lsp =
+    not (Ocaml_lsp.can_handle_typed_holes ocaml_lsp)
+  ;;
+
+  let is_valid_text_doc textdoc =
+    match TextDocument.languageId textdoc with
+    | "ocaml" | "ocaml.interface" | "reason" -> true
+    | _ -> false
+  ;;
+
+  let send_request_to_lsp client doc =
+    let uri = TextDocument.uri doc in
+    Custom_requests.send_request client Custom_requests.typedHoles uri
+  ;;
+
+  let move_to_hole range text_editor =
+    let new_selection =
+      let anchor = Range.start range in
+      let active = Range.end_ range in
+      Selection.makePositions ~anchor ~active
+    in
+    TextEditor.set_selection text_editor new_selection;
+    TextEditor.revealRange
+      text_editor
+      ~range
+      ~revealType:TextEditorRevealType.InCenterIfOutsideViewport
+      ()
+  ;;
+
+  let jump_to_hole range text_editor =
+    let open Promise.Syntax in
+    let+ _ =
+      Window.showTextDocument
+        ~document:(TextEditor.document text_editor)
+        ~preserveFocus:true
+        ()
+    in
+    move_to_hole range text_editor
+  ;;
+
+  let display_results
+        current_position
+        (results : Range.t list)
+        text_editor
+        client
+        instance
+    =
+    let open Promise.Syntax in
+    let selected_item = ref None in
+    let text_document = TextEditor.document text_editor in
+    let quickPickItems =
+      List.map results ~f:(fun res ->
+        let line = Position.line @@ Range.end_ res in
+        QuickPickItem.create
+          ~label:(Printf.sprintf "Line %d" line)
+          ~detail:
+            (Printf.sprintf
+               "%s"
+               (TextLine.text @@ TextDocument.lineAt ~line text_document))
+          ~description:(Range.json_of_t res |> Jsonoo.stringify)
```

Comment:
At the moment I don't think the vscode api gives us much room to wiggle. I have to embed the range in the description so we can fetch it later when the user selects an item.

Reply:
I think we could have custom `QuickPickItem`s, but there is also a simpler way around:

When the active item change you can get the list of all items and check what is the index of the item in that list.
Since the list is in the same order as the list of holes you can then lookup the corresponding range at that index. (you might want to turn the range list into an array)